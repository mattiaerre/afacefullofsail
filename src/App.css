.App {
  margin: auto;
  max-width: 600px;
  text-align: center;
}

.App__footer {
  font-size: small;
  margin: 4em 0 2em;
}

.Image {
  border: 1px solid black;
  margin-top: 0.5em;
}

.Legend {
  align-items: center;
  display: flex;
  flex-direction: column;
  justify-items: center;
}

.Legend__span--giveWay,
.Legend__span--standOn {
  border-style: solid;
  border-width: 0.25em;
  margin: 0.125em;
  padding: 0.25em 1em;
}

.Legend__span--giveWay {
  background-color: #ffeeee;
  border-color: #ff0000;
}

.Legend__span--standOn {
  background-color: #eeffee;
  border-color: #00ff00;
}

.List {
  align-items: baseline;
  display: flex;
  flex-direction: column;
  margin-left: 2em;
}

.Notification {
  border: 1px solid black;
  background-color: yellow;
  padding: 0.5em;
}

.Section__li--giveWay,
.Section__li--standOn {
  border-style: solid;
  border-width: 0.25em;
  margin: 1em;
  padding: 0 0 2em;
}

.Section__li--giveWay {
  background-color: #ffeeee;
  border-color: #ff0000;
}

.Section__li--standOn {
  background-color: #eeffee;
  border-color: #00ff00;
}

label {
  margin-right: 0.5em;
  text-transform: capitalize;
}

ul {
  list-style: none;
  margin: 0;
  padding: 0;
}
