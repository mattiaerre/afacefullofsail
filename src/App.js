import classnames from 'classnames';
import { useEffect, useState } from 'react';
import uuid from 'react-uuid';
import { name, version } from '../package.json';
import './App.css';
import crossing from './domain/crossing';
import Vessel from './domain/Vessel';
import {
  A_FISHING_VESSEL_WITH_NETS_IN_THE_WATER,
  BIG_SHIP,
  GIVE_WAY,
  LEEWARD,
  PORT,
  POWER_DRIVEN,
  SAILBOAT,
  STAND_ON,
  STARBOARD,
  UNKNOWN,
  WINDWARD
} from './domain/constants';

function List({ collection, name, onChange, selected }) {
  return (
    <ul>
      {collection.map((element) => {
        const id = uuid();
        return (
          <li key={element}>
            <input
              checked={element === selected}
              id={id}
              name={name}
              onChange={({ target: { value } }) => onChange(value)}
              type="radio"
              value={element}
            />
            <label htmlFor={id}>{element}</label>
          </li>
        );
      })}
    </ul>
  );
}

function Section({ name, onChange, position, tack, type }) {
  const positions = [LEEWARD, WINDWARD];

  const tacks = [PORT, STARBOARD];

  const types = [
    A_FISHING_VESSEL_WITH_NETS_IN_THE_WATER,
    BIG_SHIP,
    POWER_DRIVEN,
    SAILBOAT
  ];

  const [selectedPosition, setSelectedPosition] = useState(position);
  const [selectedTack, setSelectedTack] = useState(tack);
  const [selectedType, setSelectedType] = useState(type);

  useEffect(() => {
    const vessel = new Vessel({
      position: selectedPosition,
      tack: selectedTack,
      type: selectedType
    });

    onChange(vessel);
  }, [onChange, selectedPosition, selectedTack, selectedType]);

  return (
    <section>
      <h2>{name}</h2>
      <form>
        <h3>Type</h3>
        <List
          collection={types}
          name="type"
          onChange={setSelectedType}
          selected={selectedType}
        />
        <h3>Tack</h3>
        <List
          collection={tacks}
          name="tack"
          onChange={setSelectedTack}
          selected={selectedTack}
        />
        <h3>Position</h3>
        <List
          collection={positions}
          name="position"
          onChange={setSelectedPosition}
          selected={selectedPosition}
        />
      </form>
    </section>
  );
}

function App() {
  const [classNameA, setClassNameA] = useState(null);
  const [classNameB, setClassNameB] = useState(null);
  const [notification, setNotification] = useState(null);
  const [vesselA, setVesselA] = useState(null);
  const [vesselB, setVesselB] = useState(null);

  function makeClassName(vessel) {
    return classnames({
      'Section__li--giveWay': vessel !== null && vessel.isThe === GIVE_WAY,
      'Section__li--standOn': vessel !== null && vessel.isThe === STAND_ON
    });
  }

  useEffect(() => {
    if (vesselA !== null && vesselB !== null) {
      try {
        setNotification(null);
        crossing(vesselA, vesselB);
        setClassNameA(makeClassName(vesselA));
        setClassNameB(makeClassName(vesselB));
      } catch (error) {
        // console.log(error);
        setNotification(error.message);
        setClassNameA(makeClassName({ ...vesselA, isThe: UNKNOWN }));
        setClassNameB(makeClassName({ ...vesselB, isThe: UNKNOWN }));
      }
    }
  }, [vesselA, vesselB]);

  return (
    <article className="App">
      <p>
        "But when I raise the sail and lose those ties, anyone else in the
        cockpit inevitably gets <b>a face full of sail</b>."
      </p>
      {notification && <p className="Notification">{notification}</p>}
      <ul>
        <li className={classNameA}>
          <Section name="A" onChange={setVesselA} type={BIG_SHIP} />
        </li>
        <li className={classNameB}>
          <Section
            name="B"
            onChange={setVesselB}
            position={LEEWARD}
            tack={PORT}
            type={SAILBOAT}
          />
        </li>
      </ul>
      <footer className="App__footer">
        {new Date().getFullYear()} {name} v{version}
      </footer>
    </article>
  );
}

export default App;
