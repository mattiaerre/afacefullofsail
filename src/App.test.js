import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import App from './App';

jest.mock('react-uuid', () => () => '0000000-000-0000-000-00000000000');

jest.mock('../package.json', () => ({
  ...jest.requireActual('../package.json'),
  version: '0.0.0-jest.0'
}));

test('App snapshot', () => {
  const { container } = render(<App />);
  expect(container).toMatchSnapshot();
});

test('same vessel type', async () => {
  render(<App />);

  fireEvent.click(screen.getAllByTestId('radio')[5]); // Vessel B, type: BIG_SHIP

  expect(screen.getByTestId('notification')).toMatchSnapshot();
});
