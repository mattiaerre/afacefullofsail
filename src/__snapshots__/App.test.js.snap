// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App snapshot 1`] = `
<div>
  <article
    class="App"
  >
    <p>
      "But when I raise the sail and lose those ties, anyone else in the cockpit inevitably gets 
      <b>
        a face full of sail
      </b>
      ."
    </p>
    <h1>
      Crossing
    </h1>
    <ul>
      <li
        class="Section__li--standOn"
      >
        <section>
          <h2>
            Vessel A
          </h2>
          <form>
            <h3>
              Type
            </h3>
            <ul
              class="List"
            >
              <li>
                <input
                  id="0000000-000-0000-000-00000000000"
                  name="type"
                  type="radio"
                  value="a fishing vessel with nets in the water"
                />
                <label
                  for="0000000-000-0000-000-00000000000"
                >
                  fishing vessel
                </label>
              </li>
              <li>
                <input
                  checked=""
                  id="0000000-000-0000-000-00000000000"
                  name="type"
                  type="radio"
                  value="big ship"
                />
                <label
                  for="0000000-000-0000-000-00000000000"
                >
                  big ship
                </label>
              </li>
              <li>
                <input
                  id="0000000-000-0000-000-00000000000"
                  name="type"
                  type="radio"
                  value="power-driven"
                />
                <label
                  for="0000000-000-0000-000-00000000000"
                >
                  power-driven
                </label>
              </li>
              <li>
                <input
                  id="0000000-000-0000-000-00000000000"
                  name="type"
                  type="radio"
                  value="sailboat"
                />
                <label
                  for="0000000-000-0000-000-00000000000"
                >
                  sailboat
                </label>
              </li>
            </ul>
          </form>
        </section>
      </li>
      <li
        class="Section__li--giveWay"
      >
        <section>
          <h2>
            Vessel B
          </h2>
          <form>
            <h3>
              Type
            </h3>
            <ul
              class="List"
            >
              <li>
                <input
                  id="0000000-000-0000-000-00000000000"
                  name="type"
                  type="radio"
                  value="a fishing vessel with nets in the water"
                />
                <label
                  for="0000000-000-0000-000-00000000000"
                >
                  fishing vessel
                </label>
              </li>
              <li>
                <input
                  id="0000000-000-0000-000-00000000000"
                  name="type"
                  type="radio"
                  value="big ship"
                />
                <label
                  for="0000000-000-0000-000-00000000000"
                >
                  big ship
                </label>
              </li>
              <li>
                <input
                  id="0000000-000-0000-000-00000000000"
                  name="type"
                  type="radio"
                  value="power-driven"
                />
                <label
                  for="0000000-000-0000-000-00000000000"
                >
                  power-driven
                </label>
              </li>
              <li>
                <input
                  checked=""
                  id="0000000-000-0000-000-00000000000"
                  name="type"
                  type="radio"
                  value="sailboat"
                />
                <label
                  for="0000000-000-0000-000-00000000000"
                >
                  sailboat
                </label>
              </li>
            </ul>
            <h3>
              Tack
            </h3>
            <ul
              class="List"
            >
              <li>
                <input
                  checked=""
                  id="0000000-000-0000-000-00000000000"
                  name="tack"
                  type="radio"
                  value="port"
                />
                <label
                  for="0000000-000-0000-000-00000000000"
                >
                  port
                </label>
              </li>
              <li>
                <input
                  id="0000000-000-0000-000-00000000000"
                  name="tack"
                  type="radio"
                  value="startboard"
                />
                <label
                  for="0000000-000-0000-000-00000000000"
                >
                  startboard
                </label>
              </li>
            </ul>
            <h3>
              Position
            </h3>
            <ul
              class="List"
            >
              <li>
                <input
                  checked=""
                  id="0000000-000-0000-000-00000000000"
                  name="position"
                  type="radio"
                  value="leeward"
                />
                <label
                  for="0000000-000-0000-000-00000000000"
                >
                  leeward
                </label>
              </li>
              <li>
                <input
                  id="0000000-000-0000-000-00000000000"
                  name="position"
                  type="radio"
                  value="windward"
                />
                <label
                  for="0000000-000-0000-000-00000000000"
                >
                  windward
                </label>
              </li>
            </ul>
          </form>
        </section>
      </li>
    </ul>
    <p
      class="Legend"
    >
      <span
        class="Legend__span--giveWay"
      >
        Give-way
      </span>
      <span
        class="Legend__span--standOn"
      >
        Stand-on
      </span>
    </p>
    <footer
      class="App__footer"
    >
      2021
       
      afacefullofsail
       v
      0.0.0-jest.0
    </footer>
  </article>
</div>
`;
